### PR TITLE
Skip CRIS import record updates if a key column is missing a value

### PR DESCRIPTION
**Pull request by mddilley**
## Associated issues
See https://austininnovation.slack.com/archives/CM9MK950S/p1713190489691989

TLDR: Older records in the person and primary person tables that are missing the `prsn_occpnt_pos_id` are breaking the SQL that we use to query and update records that receive updates from CRIS. This is one of the columns that [we use to check if a person row is unique](https://github.com/cityofaustin/atd-vz-data/blob/5b69bab6f35677c815b412427674da71809b70b7/atd-etl/cris_import/lib/mappings.py#L23-L36).

## Testing
**URL to test:** 
Local

**Steps to test:**
1. Ask me for a zip file that I created that contains csvs with only one crash and its related person, primary person, and unit records that can replicate the original bug 🪄🪱 You can see our current values for the people records in the test csv with the following query (which some of these appear to be duplicates):
```sql
SELECT
	atc.crash_id,
	atc.crash_date,
	atp.prsn_nbr,
	atp.unit_nbr,
	atp.prsn_type_id,
	atp.prsn_occpnt_pos_id
FROM
	atd_txdot_person atp
	LEFT JOIN atd_txdot_crashes atc ON atp.crash_id = atc.crash_id
WHERE
	atc.crash_id = 14003860;
```
2. See readme for local testing instructions
3. Run the import on the zip and you should see the person records that are missing the `prsn_occpnt_pos_id` column value logged as skipped

**Releasing this:**
We should see the [production image of the CR3 import](https://hub.docker.com/layers/atddocker/vz-cris-import/production/images/sha256-9e346d0548dc7a69fbbea4ffc524429da7015596ce4c40eaa2ff7add98ed30c3?context=repo) update after merging this to `master`. [GH action code](https://github.com/cityofaustin/atd-vz-data/blob/5b69bab6f35677c815b412427674da71809b70b7/.github/workflows/build_docker_images.yml#L73-L80)

---
#### Ship list
- [ ] Check migrations for any conflicts with latest migrations in master branch
- [ ] Confirm Hasura role permissions for necessary access
- [ ] Code reviewed
- [ ] Product manager approved

### CODE REVIEW

File: atd-etl/cris_import/cris_import.py [hunk 1]
```diff
@@ -612,7 +612,23 @@ def align_records(map_state):
...
 
             # iterate over each imported record and determine correct action
+            should_skip_update = False
+
             for source in imported_records:
+                # Check unique key columns to make sure they all have a value
+                for key_column in key_columns:
+                    key_column_value = source[key_column]
+                    if key_column_value == None:
+                        print("\nSkipping because unique key column is missing")
```
Comment (mddilley) [inline atd-etl/cris_import/cris_import.py:622]: Super open to whatever amount of logging we want here. 🙏 I used this because it works good for testing.

File: atd-etl/cris_import/cris_import.py [hunk 2]
```diff
@@ -612,7 +612,23 @@ def align_records(map_state):
             input_column_names = util.get_input_column_names(pg, map_state["import_schema"], table, target_columns)
 
             # iterate over each imported record and determine correct action
+            should_skip_update = False
+
             for source in imported_records:
+                # Check unique key columns to make sure they all have a value
```
Comment (mddilley) [inline atd-etl/cris_import/cris_import.py:618]: We discussed specifically looking at people tables, but this should protect us from updating records in other tables when we don't have all the info we need to make sure we are updating the correct record.

File: atd-etl/cris_import/README.md [hunk 3]
```diff
@@ -12,3 +12,12 @@ The contents of this directory define a Docker image that can be used in an Airf
...
 The CRIS import usually works by pulling down a zip archive from the SFTP endpoint. However, during development, it's much easier if you can have it use a zip file that you have locally on your machine instead. This can be accomplished by putting a zip file (still encrypted and using "CRIS extract" password) in a folder named `atd-etl/cris_import/development_extracts/`. Create the directory if needed. If there are no zip files in that folder, the program will automatically revert to inspecting the SFTP endpoint.
+
+### Local testing
+
+Make a copy of `env-template` and name it `env`. Fill in the values using the 1Password Connect Server secrets (see entries titled `Endpoint for 1Password Connect Server API` and `Vault ID of API Accessible Secrets vault`) and your personal access token.
+
+Drop a CRIS extract zip file into your development folder as described above, and run the import script:
+```bash
+docker compose run cris-import
+```
```
Comment (johnclary) [inline atd-etl/cris_import/README.md:23]: 🙌

File: atd-etl/cris_import/cris_import.py (hunk 1 above)
Comment (johnclary) [reply atd-etl/cris_import/cris_import.py:622]: this looks good to me. maybe add another `print` here that prints the entire `source`, so that we have some hope of tracking this record down?

File: atd-etl/cris_import/cris_import.py [hunk 4]
```diff
@@ -612,7 +612,23 @@ def align_records(map_state):
...
+                    if key_column_value == None:
+                        print("\nSkipping because unique key column is missing")    
+                        print(f"Table: {table}")
+                        print(f"Missing key column: {key_column}")
+                        should_skip_update = True
+
+                # If we are missing a column that uniquely identifies the record, we should skip the update
+                if should_skip_update:
+                    for key_column in key_columns:
+                        print(f"{key_column}: {source[key_column]}")
```
Comment (johnclary) [inline atd-etl/cris_import/cris_import.py:630]: i think you can remove this `print` since you have the more verbose one above 👍

File: atd-etl/cris_import/env-template
Comment (johnclary) [inline atd-etl/cris_import/env-template:1]: thank you!!!

File: atd-etl/cris_import/cris_import.py [hunk 5]
```diff
@@ -612,7 +612,23 @@ def align_records(map_state):
             input_column_names = util.get_input_column_names(pg, map_state["import_schema"], table, target_columns)
 
             # iterate over each imported record and determine correct action
+            should_skip_update = False
+
             for source in imported_records:
+                # Check unique key columns to make sure they all have a value
+                for key_column in key_columns:
+                    key_column_value = source[key_column]
+                    if key_column_value == None:
```
Comment (johnclary) [inline atd-etl/cris_import/cris_import.py:621]: super minor, but i always see this test written as `is None` instead of `== None`. there is some arcane pythonic reason for it that i am too lazy to look up 😅

**Review by johnclary — APPROVED** [review]
i have not tested, but this looks fantastic too me. thank you @mddilley!! 🚢 🚢 🚢 🚢 🚢

File: atd-etl/cris_import/cris_import.py (hunk 5 above)
Comment (mddilley) [reply atd-etl/cris_import/cris_import.py:621]: sounds great 🐍 I'll push this up along with the other minor ones above. Thanks! 🙏

**Review by frankhereford — APPROVED** [review]
## Before the fix:
```
Traceback (most recent call last):
  File "/app/./cris_import.py", line 831, in <module>
    main()
  File "/app/./cris_import.py", line 108, in main
    align_records_token = align_records(typed_token)
  File "/app/./cris_import.py", line 622, in align_records
    if util.fetch_target_record(pg, output_map, table, record_key_sql):
  File "/app/lib/sql.py", line 169, in fetch_target_record
    cursor.execute(sql)
  File "/usr/local/lib/python3.10/dist-packages/psycopg2/extras.py", line 236, in execute
    return super().execute(query, vars)
psycopg2.errors.UndefinedColumn: column "none" does not exist
LINE 5: ...id = 2 and public.atd_txdot_person.prsn_occpnt_pos_id = None
```

## After the fix
```
Finding updated records

Skipping because unique key column is missing
Table: person
Missing key column: prsn_occpnt_pos_id
crash_id: 1*******0
unit_nbr: 2
prsn_nbr: 3
prsn_type_id: 2
prsn_occpnt_pos_id: None

Skipping because unique key column is missing
Table: person
Missing key column: prsn_occpnt_pos_id
crash_id: 1********0
unit_nbr: 2
prsn_nbr: 4
prsn_type_id: 2
prsn_occpnt_pos_id: None
root@85b604a1dcaa:/app# 
```

Perfect! 🚢🚢🚢🚢

File: atd-etl/cris_import/cris_import.py (hunk 1 above)
Comment (johnclary) [reply atd-etl/cris_import/cris_import.py:622]: Mike pointed out that we don't want to print people records to our logs. i retract this suggestion 🙏

Comment (mddilley) [reply atd-etl/cris_import/cris_import.py:622]: chatted with John about this and decided to keep as is so we don't print out sensitive people data. 🔒

File: atd-etl/cris_import/cris_import.py (hunk 4 above)
Comment (mddilley) [reply atd-etl/cris_import/cris_import.py:630]: since we aren't printing the source record, i'll keep this here since it will gives us the details to track down which record was skipped no matter what table it is in.
